File: scripts/build-and-publish.ts
/* eslint-disable */

import fs from 'fs';
import path from 'path';
import { spawnSync } from 'child_process';

type ChangedPlugin = {
  name: string;
};

type PackageInfo = {
  name: string;
  version: string;
};

type OutputEntry = {
  disabled: boolean;
  name: string;
  version: string;
  integrity: string;
};

const PLUGIN_DIR = path.resolve(__dirname, '../plugins');
const OUTPUT_FILE = path.resolve(__dirname, '../plugins-output.yaml');
const CHANGED_JSON = path.resolve(__dirname, '../changed.json');

/**
 * Lê os plugins modificados a partir de `changed.json`
 */
function readChangedPlugins(): string[] {
  if (!fs.existsSync(CHANGED_JSON)) return [];
  const content = fs.readFileSync(CHANGED_JSON, 'utf-8');
  const parsed = JSON.parse(content) as ChangedPlugin[];
  return parsed.map(pkg => pkg.name);
}

/**
 * Lê name e version do package.json do plugin
 */
function getPackageInfo(pluginPath: string): PackageInfo {
  const pkgPath = path.join(pluginPath, 'package.json');
  const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'));
  return { name: pkg.name, version: pkg.version };
}

/**
 * Retorna o hash de integridade de um pacote após rodar `npm pack`
 */
function getIntegrityFromTarball(
  pluginPath: string,
  pkgName: string,
): string | null {
  const packResult = spawnSync('npm', ['pack', '--json'], {
    cwd: pluginPath,
    encoding: 'utf-8', // stdout será string
  });

  if (packResult.status !== 0) {
    console.error(`❌ Erro ao gerar pacote para ${pkgName}`);
    console.error(packResult.stderr);
    return null;
  }

  const outputJson = JSON.parse(packResult.stdout)[0];
  return outputJson.integrity as string;
}

/**
 * Publica o plugin no NPM via dist-dynamic/
 */
function publishPlugin(pluginPath: string, pkgName: string): void {
  console.log(`📦 Publicando ${pkgName}...`);

  const distDynamic = path.join(pluginPath, 'dist-dynamic');
  const publish = spawnSync('npm', ['publish', '--access', 'public'], {
    cwd: distDynamic,
    stdio: 'inherit',
    env: {
      ...process.env,
      NODE_AUTH_TOKEN: process.env.NODE_AUTH_TOKEN, // necessário no CI
    },
  });

  if (publish.status !== 0) {
    console.error(`❌ Falha ao publicar ${pkgName}`);
    process.exit(1);
  }
}

/**
 * Roda build-dynamic-plugin via yarn
 */
function buildDynamic(pluginPath: string, pkgName: string): void {
  console.log(`🛠️  Buildando ${pkgName}...`);
  const result = spawnSync('yarn', ['build-dynamic-plugin'], {
    cwd: pluginPath,
    stdio: 'inherit',
  });

  if (result.status !== 0) {
    console.error(`❌ Falha ao buildar ${pkgName}`);
    process.exit(1);
  }
}

/**
 * Escreve no arquivo plugins-output.yaml
 */
function writeOutputEntry(entry: OutputEntry): void {
  const content = `- disabled: ${entry.disabled}
  package: ${entry.name}@${entry.version}
  integrity: ${entry.integrity}
`;
  fs.appendFileSync(OUTPUT_FILE, content);
}

/**
 * Runner principal
 */
function run(): void {
  console.log('🚀 Iniciando processo de build e publicação...');

  if (fs.existsSync(OUTPUT_FILE)) fs.unlinkSync(OUTPUT_FILE);

  const changed = readChangedPlugins();

  const pluginDirs = fs.readdirSync(PLUGIN_DIR).filter(d => {
    const fullPath = path.join(PLUGIN_DIR, d);
    return fs.statSync(fullPath).isDirectory();
  });

  for (const dir of pluginDirs) {
    const fullPath = path.join(PLUGIN_DIR, dir);
    const { name, version } = getPackageInfo(fullPath);
    const wasChanged = changed.includes(name);

    if (wasChanged) {
      buildDynamic(fullPath, name);
      publishPlugin(fullPath, name);
    }

    const integrity = getIntegrityFromTarball(fullPath, name);
    if (!integrity) continue;

    writeOutputEntry({
      disabled: !wasChanged,
      name,
      version,
      integrity,
    });
  }

  console.log(`✅ Processo finalizado! Output em: ${OUTPUT_FILE}`);
}

run();

File: scripts/build-and-publish.py
#!/usr/bin/env python3

import json
import os
import subprocess
import sys
import yaml
from pathlib import Path


try:
    from packaging.version import Version
except ImportError:
    print("❌ Missing required dependency: packaging")
    print("💡 Run: pip install packaging")
    sys.exit(1)

ROOT = Path(__file__).resolve().parent.parent
PLUGIN_DIR = ROOT / "plugins"
CHANGED_JSON = ROOT / "changed.json"
OUTPUT_FILE = ROOT / "plugins-output.yaml"

def read_changed_plugins():
    if not CHANGED_JSON.exists():
        return []
    with open(CHANGED_JSON) as f:
        data = json.load(f)
    return [pkg["name"] for pkg in data]

def get_package_info(plugin_path):
    pkg_path = plugin_path / "package.json"
    with open(pkg_path) as f:
        pkg = json.load(f)
    return pkg["name"], pkg["version"]

def bump_patch(version_str):
    v = Version(version_str)
    return f"{v.major}.{v.minor}.{v.micro + 1}"

def bump_version(plugin_path, old_version):
    new_version = bump_patch(old_version)
    pkg_path = plugin_path / "package.json"
    with open(pkg_path) as f:
        data = json.load(f)
    data["version"] = new_version
    with open(pkg_path, "w") as f:
        json.dump(data, f, indent=2)
    print(f"🔧 Bumped version from {old_version} → {new_version}")
    return new_version

def run_command(command, cwd):
    print(f"▶️ Running: {' '.join(command)} in {cwd}")
    result = subprocess.run(command, cwd=cwd)
    if result.returncode != 0:
        raise Exception(f"❌ Command failed: {' '.join(command)}")

def build_dynamic(plugin_path, name):
    print(f"🔨 Building {name}...")
    if (plugin_path / "tsconfig.json").exists():
        run_command(["yarn", "tsc"], cwd=plugin_path)
    run_command(["yarn", "build-dynamic-plugin"], cwd=plugin_path)

def publish_plugin(plugin_path, name):
    print(f"📦 Publishing {name}...")
    dist_path = plugin_path / "dist-dynamic"
    run_command(["npm", "publish", "--access", "restricted"], cwd=dist_path)

def get_integrity(plugin_path, name):
    print(f"📦 Packing {name} to get integrity hash...")
    result = subprocess.run(
        ["npm", "pack", "--json"],
        cwd=plugin_path,
        capture_output=True,
        text=True
    )
    if result.returncode != 0:
        raise Exception(f"❌ Failed to generate package for {name}")
    output = json.loads(result.stdout)[0]
    return output["integrity"]

def main():
    print("🚀 Starting build and publish process...\n")

    if OUTPUT_FILE.exists():
        OUTPUT_FILE.unlink()

    changed = read_changed_plugins()
    plugins = sorted(p for p in PLUGIN_DIR.iterdir() if p.is_dir())
    output = []

    for plugin_path in plugins:
        try:
            name, version = get_package_info(plugin_path)
            changed_flag = name in changed

            if changed_flag:
                new_version = bump_version(plugin_path, version)
                build_dynamic(plugin_path, name)
                publish_plugin(plugin_path, name)
                version = new_version

            integrity = get_integrity(plugin_path, name)

        except Exception as e:
            print(f"❌ Failed processing {plugin_path.name}: {e}")
            integrity = "unknown"
            version = version if 'version' in locals() else "0.0.0"
            name = name if 'name' in locals() else plugin_path.name

        output.append({
            "disabled": name not in changed,
            "package": f"{name}@{version}",
            "integrity": integrity
        })

    with open(OUTPUT_FILE, "w") as f:
        yaml.dump({"plugins": output}, f, default_flow_style=False)

    print(f"\n✅ Finished! Output written to: {OUTPUT_FILE}")

if __name__ == "__main__":
    main()
